contracts/zero-ex/contracts/src/features/UniswapV2ForkFeature.sol

```diff
@@ -0,0 +1,200 @@
+            tokens[0],
+            msg.sender,
+            address(_pairFor(tokens[0], tokens[1])),
+            sellAmount
+        );
+        address recipient = unwrapWeth ? address(this) : msg.sender;
+        uint[] memory amounts = _swap(tokens, sellAmount, recipient);
+        buyAmount = amounts[amounts.length - 1];
+        require(buyAmount >= minBuyAmount, 'UniswapV2ForkFeature/INSUFFICIENT_OUTPUT_AMOUNT');
+
+        if (unwrapWeth) {
+            IEtherTokenV06(WETH).withdraw(buyAmount);
+            (bool sent, ) = msg.sender.call{value: buyAmount}("");
+            require(sent, "UniswapV2ForkFeature/FAILED_TO_SEND_ETHER");
+        }
+    }
+
+    function _sortTokens(
+        IERC20TokenV06 tokenA,
+        IERC20TokenV06 tokenB
+    )
+        internal
+        pure
+        returns (IERC20TokenV06 token0, IERC20TokenV06 token1)
+    {
+        require(tokenA != tokenB, 'UniswapV2ForkFeature/IDENTICAL_ADDRESSES');
+        (token0, token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
+        require(token0 != IERC20TokenV06(0), 'UniswapV2ForkFeature/ZERO_ADDRESS');
+    }
+
+    function _getAmountOut(
+        uint amountIn,
+        uint reserveIn,
+        uint reserveOut
+    )
+        internal
+        pure
+        returns (uint amountOut)
+    {
+        require(amountIn > 0, 'UniswapV2ForkFeature/INSUFFICIENT_INPUT_AMOUNT');
+        require(reserveIn > 0 && reserveOut > 0, 'UniswapV2ForkFeature/INSUFFICIENT_LIQUIDITY');
+        uint amountInWithFee = amountIn.safeMul(997);
+        uint numerator = amountInWithFee.safeMul(reserveOut);
+        uint denominator = reserveIn.safeMul(1000).safeMul(amountInWithFee);
+        amountOut = numerator / denominator;
+    }
+
+    function _pairFor(
+        IERC20TokenV06 tokenA,
+        IERC20TokenV06 tokenB
+    )
+        internal
+        view
+        returns (IUniswapV2ForkPair)
+    {
+        (IERC20TokenV06 token0, IERC20TokenV06 token1) = _sortTokens(tokenA, tokenB);
+        return IUniswapV2ForkPair(address(uint(keccak256(abi.encodePacked(
+            hex'ff',
+            UNI_FACTORY_ADDRESS,
+            keccak256(abi.encodePacked(address(token0), address(token1))),
+            UNI_POOL_INIT_CODE_HASH
+        )))));
+    }
+
+    function _swap(
+        IERC20TokenV06[] memory tokens,
+        uint256 sellAmount,
+        address _to
+    )
+        internal
+        returns (uint[] memory amounts)
+    {
+        require(tokens.length >= 2, 'UniswapV2ForkFeature/INVALID_PATH');
+
+        amounts = new uint[](tokens.length);
+        amounts[0] = sellAmount;
+
+        for (uint i; i < tokens.length - 1; i++) {
+            (IERC20TokenV06 tokenA, IERC20TokenV06 tokenB, IERC20TokenV06 tokenC) =
+                (tokens[i], tokens[i + 1], tokens[i + 2]);
+
+            address to = i < tokens.length - 2 ? address(_pairFor(tokenB, tokenC)) : _to;
+            amounts[i + 1] = _swapInternal(amounts[i], tokenA, tokenB, to);
+        }
+    }
```

Comment:
something like
```suggestion
    function _swap(
        IERC20TokenV06[] memory tokens,
        uint256 sellAmount,
        address recipient
    )
        internal
        returns (uint[] memory amounts)
    {
        require(tokens.length >= 2, 'UniswapV2ForkFeature/INVALID_PATH');
    
        IUniswapV2ForkPair currentPair = _pairFor(
            _normalizeToken(tokens[0]),
            _normalizeToken(tokens[1])
        );

        if (tokens[0] == ETH_ADDRESS) {
            _transferERC20Tokens(WETH, currentPair, sellAmount);
        } else {
            _transferERC20TokensFrom(WETH, msg.sender, currentPair, sellAmount);
        }

        for (uint256 i; i < tokens.length - 1; i++) {
            IUniswapV2ForkPair nextPair;
            if (i < tokens.length - 2) {
                nextPair = _pairFor(
                    _normalizeToken(tokens[i + 1]),
                    _normalizeToken(tokens[i + 2])
                );
            }
            sellAmount = _swapInternal(
                _normalizeToken(tokens[i]),
                _normalizeToken(tokens[i + 1]),
                currentPair,
                sellAmount, // sell last bought amount
                address(nextPair) ? address(nextPair) : recipient
            );
            currentPair = nextPair;
        }
        return amount;
    }
```

where `_normalizeToken()` just converts ETH_ADDRESS to WETH

Reply:
do we need to do `_normalizeToken`  every time, shouldn't we revert once we see a ETH in the middle of the path? I pushed the change without doing so, PTAL, thx!

Reply:
ETH/WETH is often in the middle of a path. A->WETH->B is a common hop. We just treat intermediate ETHs as WETH

Reply:
i guess here is what i confused about, I understand that we could have intermediate WETH, but how we get an intermediate ETH on Uni V2 forks, which they doesnt support? Is it from our centralized service?

Reply:
yeah we won't ever get ETH from a uniswap swap, only WETH, so whether we encounter `WETH` or `ETH_ADDDRESS` in `tokens` we will treat it as `WETH` here.

Reply:
sorry, I guess I didnt ask it right, how can we ever encounter `ETH_ADDDRESS` in the **MIDDLE** of `tokens`? Shouldn't we expect the `tokens` is prepared so that it does not include ETH in the middle prior to calling Univ2 feature?

Reply:
Yes, it won't ever happen, but the function will treat it as WETH for simplicity.